File: wetty-chat-mobile/src/api/savedMessages.ts
import type { AxiosResponse } from 'axios';
import type { MentionInfo, MessageResponse, UserGroupTagInfo } from './messages';
import apiClient from './client';

export interface SavedAttachmentSnapshot {
  id: string;
  externalReference: string;
  url: string;
  kind: string;
  size: number;
  fileName: string;
  width?: number | null;
  height?: number | null;
  order: number;
}

export interface SavedStickerSnapshot {
  id: string;
  emoji: string;
  name?: string | null;
  mediaUrl: string;
  mediaContentType: string;
}

export interface SavedSenderSnapshot {
  uid: number;
  name: string | null;
  avatarUrl?: string | null;
  gender: number;
  userGroup?: UserGroupTagInfo | null;
}

export interface SavedChatSnapshot {
  id: string;
  name: string | null;
  avatarUrl?: string | null;
}

export interface SavedMessageResponse {
  id: string;
  originalChatId: string;
  originalThreadRootId: string | null;
  originalMessageId: string;
  originalReplyToMessageId: string | null;
  originalSenderUid: number;
  originalCreatedAt: string;
  savedAt: string;
  message: string | null;
  messageType: MessageResponse['messageType'];
  attachments: SavedAttachmentSnapshot[];
  sticker?: SavedStickerSnapshot | null;
  mentions: MentionInfo[];
  sender: SavedSenderSnapshot;
  chat: SavedChatSnapshot;
  canLocateContext: boolean;
}

export interface ListSavedMessagesResponse {
  savedMessages: SavedMessageResponse[];
  nextCursor: string | null;
}

interface ListSavedMessagesParams {
  limit?: number;
  before?: string | null;
}

function compactListParams(params?: ListSavedMessagesParams): Record<string, string | number> {
  const query: Record<string, string | number> = {};
  if (params?.limit != null) query.limit = params.limit;
  if (params?.before != null) query.before = params.before;
  return query;
}

export function saveMessage(messageId: string): Promise<AxiosResponse<SavedMessageResponse>> {
  return apiClient.put(`/saved-messages/${messageId}`);
}

export function deleteSavedMessage(savedMessageId: string): Promise<AxiosResponse<void>> {
  return apiClient.delete(`/saved-messages/${savedMessageId}`);
}

export function deleteSavedMessageByOriginal(messageId: string): Promise<AxiosResponse<void>> {
  return apiClient.delete(`/saved-messages/by-message/${messageId}`);
}

export function listSavedMessages(params?: ListSavedMessagesParams): Promise<AxiosResponse<ListSavedMessagesResponse>> {
  return apiClient.get('/saved-messages', { params: compactListParams(params) });
}

export function listChatSavedMessages(
  chatId: string,
  params?: ListSavedMessagesParams,
): Promise<AxiosResponse<ListSavedMessagesResponse>> {
  return apiClient.get(`/chats/${chatId}/saved-messages`, { params: compactListParams(params) });
}

File: backend/src/dto/saved_messages.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::{
    dto::{messages::MentionInfo, users::UserGroupTagInfo},
    models::MessageType,
};

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SavedAttachmentSnapshot {
    #[serde(with = "crate::serde_i64_string")]
    #[schema(value_type = String)]
    pub id: i64,
    pub external_reference: String,
    pub url: String,
    pub kind: String,
    pub size: i64,
    pub file_name: String,
    pub width: Option<i32>,
    pub height: Option<i32>,
    pub order: i16,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SavedStickerSnapshot {
    #[serde(with = "crate::serde_i64_string")]
    #[schema(value_type = String)]
    pub id: i64,
    pub emoji: String,
    pub name: Option<String>,
    pub media_url: String,
    pub media_content_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SavedSenderSnapshot {
    pub uid: i32,
    pub name: Option<String>,
    pub avatar_url: Option<String>,
    pub gender: i16,
    #[serde(default)]
    pub user_group: Option<UserGroupTagInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SavedChatSnapshot {
    #[serde(with = "crate::serde_i64_string")]
    #[schema(value_type = String)]
    pub id: i64,
    pub name: String,
    pub avatar_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SavedMessageResponse {
    #[serde(with = "crate::serde_i64_string")]
    #[schema(value_type = String)]
    pub id: i64,
    #[serde(with = "crate::serde_i64_string")]
    #[schema(value_type = String)]
    pub original_chat_id: i64,
    #[serde(with = "crate::serde_i64_string::opt")]
    #[schema(value_type = Option<String>)]
    pub original_thread_root_id: Option<i64>,
    #[serde(with = "crate::serde_i64_string")]
    #[schema(value_type = String)]
    pub original_message_id: i64,
    #[serde(with = "crate::serde_i64_string::opt")]
    #[schema(value_type = Option<String>)]
    pub original_reply_to_message_id: Option<i64>,
    pub original_sender_uid: i32,
    pub original_created_at: DateTime<Utc>,
    pub saved_at: DateTime<Utc>,
    pub message: Option<String>,
    pub message_type: MessageType,
    pub attachments: Vec<SavedAttachmentSnapshot>,
    pub sticker: Option<SavedStickerSnapshot>,
    pub mentions: Vec<MentionInfo>,
    pub sender: SavedSenderSnapshot,
    pub chat: SavedChatSnapshot,
    pub can_locate_context: bool,
}

#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ListSavedMessagesResponse {
    pub saved_messages: Vec<SavedMessageResponse>,
    #[serde(with = "crate::serde_i64_string::opt")]
    #[schema(value_type = Option<String>)]
    pub next_cursor: Option<i64>,
}
